feat: Responsive Styles Support (#1779)
Workday/canvas-kit: modules/react/common/stories/Responsive.stories.mdx

import {Meta, ArgsTable} from '@storybook/addon-docs';
import {Box} from '@workday/canvas-kit-react/layout';
import {useResizeObserver, useResponsiveContainerStyles} from '@workday/canvas-kit-react/common';
import {ResponsiveContainer} from './examples/ResponsiveStyles';
import {ResponsiveViewport} from './examples/ResponsiveViewport';

<Meta title="Hooks and Utilities/Responsive Styling" />

# Responsive Styling

> Responsive styling is an approach to web development that adjusts layout and appearance based on the
> device's viewport size or an element's container size.
> 
> ## What is Responsive Styling?
> 
> Responsive Styling it is a term used to describe an approach to web design or a set of best
> practices, used to create a layout that can _respond_ to the device being used to view the content.

Reviewer: We don't need that paragraph under the H1 since it'll just be stripped out on the site. Moving it underneath `## What is Responsive Styling?`.
```suggestion
## What is Responsive Styling?

Responsive styling is an approach to web development that adjusts layout and appearance based on the
device's viewport size or an element's container size.
```

Read more about the history of
[Responsive Design](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Responsive_Design).

> ### Container vs Viewport
> 
> There are two ways we will look at responsive styling below. They are container-based and
> viewport-based responsive styling.
> 
> > Using container-based responsive styling, we will measure the size of the container the element is
> > in and apply styles based on the size of that container.
> 
> > Using viewport-based responsive styling, we will measure the size of the user's viewport and apply
> > styles based on the size of that viewport.
> 
> If you are measuring a user's viewport size, then you should be using viewport-based responsive
> styling.
> 
> If you only want to measure the size of a container, then you should be using container-based
> responsive styling.
> 
> > **Note**: A container's size can be affected by surrounding containers or the viewport size.

Reviewer: I updated this content to remove some of the repetitive language and to segue into the next topic of breakpoints.
```suggestion
### Container-based vs. Viewport-based Responsive Styling

There are two approaches to responsive styling:

1. With **container-based** styling, we measure the size of the container an element is in and apply
   styles based on the size of that container. Note that a container's size can be affected by
   surrounding containers or the viewport size.
2. With **viewport-based** styling, we measure the size of the user's viewport and apply styles
   based on the size of that viewport.

Both approaches use breakpoints.
```

### Canvas Breakpoints

> In Canvas Kit, there are a set of breakpoints:
> 
> | Zero  | Small   | Medium  | Large    | Extra Large |
> | ----- | ------- | ------- | -------- | ----------- |
> | > 0px | > 320px | > 768px | > 1024px | > 1440px    |
> 
> > Zero: This is applied from 0 - 319px
> 
> > Small: This is applied from 320px - 767px
> 
> > Medium: This is applied from 768px - 1023px
> 
> > Large: This is applied from 1024px - 1439px
> 
> > Extra Large: This is applied from 1440px and up
> 
> These are the breakpoints we will use when we apply styles based on the size of a container or
> viewport.

Reviewer: I updated this section to remove more duplicate content and to provide an explanation of breakpoints.
```suggestion
Breakpoints are the points at which the styles will change to provide an optimal user experience.
Think of a container element's or viewport's width as a range of possible sizes broken down into
buckets (e.g., `0-319px`, `320-767px`, `768-1023px`, etc).

Canvas uses the following breakpoints for both container-based and viewport-based responsive
styling:

- `zero`: styles applied from `0px` up until (but not including) `320px`
- `s`: styles applied from `320px` up until (but not including) `768px`
- `m`: styles applied from `768px` up until (but not including) `1024px`
- `l`: styles applied from `1024px` up until (but not including) `1440px`
- `xl`: styles applied from `1440px` and up
```

change to match the size shown in the dropdown.

<ExampleCodeBlock code={ResponsiveContainer} />

This example uses the following hooks:

- `useResizeObserver` provides the width (and height) of the container.

- `useResponsiveContainerStyles` allows you to create container-based responsive styles with
  objects. This hook accepts three arguments: style objects, container width, and theme (optional).
  Each style object accepts keys corresponding to the following sizes: `zero`, `s`, `m`, `l` and
  `xl`. These sizes represent [Canvas breakpoints](#canvas-breakpoints). The sizes act like
  `min-width`. For example, if you want to apply styles from `medium` and up, then you would write
  those styles under `m`.

## Viewport-based Responsive Styling

The viewport is the user's visible area of a web page (varies with screen size).

We can use media queries to run a series of tests that will measure the screen size or resolution
and apply CSS styles specifically to an element.

```css
@media screen and (min-width: 800px) {
  .container {
    margin: 1em 2em;
  }
}
```

In the interactive example below, styles are applied based on the size of the user's viewport using

> our breakpoint utilities: `up()`, `down()`, `between()` and `only()`. Note the use of `up()` and
> `down()` near the top of the example code. Adjust the width of your browser to see the layout adjust
> to the available width.

Reviewer: I'm removing the parentheses from `up()` and `down()`. We don't typically use parentheses when referencing functions in our documentation though I'm wondering if we should to make it more obvious we're referring to a function. Something to think about for the future.

```suggestion
two of our breakpoint utilities: `up` and `down`. Adjust the width of your browser to see the layout
adjust to the available width.
```

<ExampleCodeBlock code={ResponsiveViewport} />

### Breakpoint Utilities

> We have four utilities to apply viewport-based responsive styling: `up()`, `down()`, `between()`,
> and `only()`. Each has a specific role in applying viewport-based responsive styles.

Reviewer: ```suggestion
We have four utilities to apply viewport-based responsive styling: `up`, `down`, `between`, and
`only`. Each has a specific role in applying viewport-based responsive styles.
```

> `up()` - Returns a media query above the `min-width` for the range of a given breakpoint. Given a
> `start` breakpoint key `("zero", "s", "m", "l", "xl")`, this function returns a media query (string)
> using a `min-width`. See example below.

Reviewer: Let's make `up` into a heading to visually set it apart.
```suggestion
#### `up`

`up` returns a media query above the `min-width` for the range of a given breakpoint. Given a
`start` breakpoint key `("zero", "s", "m", "l", "xl")`, this function returns a media query (string)
using a `min-width`.
```

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';
const theme = useTheme();
const {up} = theme.canvas.breakpoints;
const mediaQuery = up('l'); // Returns '@media (min-width: 1024px)'
const styles = {
  [mediaQuery]: {
    padding: space.m,
  },
};
```

> `down()` - Returns a media query below the `max-width` for the range of a given breakpoint. Given an

Reviewer: Same deal, let's make `down` into a heading.
```suggestion
#### `down`

`down` returns a media query below the `max-width` for the range of a given breakpoint. Given an
```

`end` breakpoint key `("zero", "s", "m", "l", "xl")`, this function returns a media query (string)
using a `max-width`.

> **Note**: This function subtracts `0.5px` from the next breakpoint value to prevent collisions.
> For example, `breakpoints.values.s`, has a `min-width` of `320px`, and the `max-width` is
> `767.5px`). If the `xl` breakpoint is provided, this function returns a media query with only a

> > `min-width` of `0`, as seen in the second example below. See example below.

Reviewer: ```suggestion
> `min-width` of `0`, as seen in the second example below.
```

Author: Yeah repeated myself there

```ts
import { useTheme } from '@workday/canvas-kit-react/common';
import { space } from '@workday/canvas-kit-react/tokens';
 *
const theme = useTheme();
const { down } = theme.canvas.breakpoints;
const mediaQuery = down('m'); // Returns '@media (max-width: 1023.5px)'
const styles = {
  [mediaQuery]: {
    padding: space.m,
  }
};
```

This example uses the `xl` breakpoint and only adds a `min-width` of `0` to the media query.

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';
const theme = useTheme();
const {down} = theme.canvas.breakpoints;
const mediaQuery = down('xl'); // Returns '@media (min-width: 0)'
const styles = {
  [mediaQuery]: {
    padding: space.m,
  },
};
```

> `between()` - Returns a media query between two given breakpoints. Given `start` and `end`
> breakpoint keys ("zero", "s", "m", "l", "xl"), this function returns a media query (string) using a
> min-width and max-width.

Reviewer: Same deal, let's make `between` into a heading. Added missing backticks as well.
```suggestion
#### `between`

`between` returns a media query between two given breakpoints. Given `start` and `end` breakpoint
keys `("zero", "s", "m", "l", "xl")`, this function returns a media query using a `min-width` and
`max-width`.
```

> **Note**: This function subtracts `0.5px` from the next breakpoint value to prevent collisions.
> For example, `breakpoints.values.s`, has a `min-width` of `320px`, and the `max-width` is
> `767.5px`).

If the `xl` breakpoint is provided, this function returns a media query with only a `min-width`, as
seen in the second example below.

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';
const theme = useTheme();
const {between} = theme.canvas.breakpoints;
// Returns '@media (min-width: 320px) and (max-width: 1023.5px)'
const mediaQuery = between('s', 'm');
const styles = {
  [mediaQuery]: {
    padding: space.s,
  },
};
```

This example uses `xl` as the `end` breakpoint and only adds a min-width to the media query.

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';
const theme = useTheme();
const {between} = theme.canvas.breakpoints;
const mediaQuery = between('m', 'xl'); // Returns '@media (min-width: 768px)'
const styles = {
  [mediaQuery]: {
    padding: space.s,
  },
};
```

> `only()` - Returns a media query with a `min-width` and `max-width` for a given breakpoint. Given a
> breakpoint key ("zero", "s", "m", "l", "xl"), this function returns a media query (string) using a

Reviewer: Same deal, let's make `only` into a heading. Added missing backticks as well.
```suggestion
#### `only`

`only` returns a media query with a `min-width` and `max-width` for a given breakpoint. Given a
breakpoint key `("zero", "s", "m", "l", "xl")`, this function returns a media query using a
```

import {Meta, ArgsTable} from '@storybook/addon-docs';
import {Box} from '@workday/canvas-kit-react/layout';
import {useResizeObserver, useResponsiveContainerStyles} from '@workday/canvas-kit-react/common';
import {ResponsiveContainer} from './examples/ResponsiveStyles';
import {ResponsiveViewport} from './examples/ResponsiveViewport';

<Meta title="Hooks and Utilities/Responsive Styling" />

# Responsive Styling

> Responsive styling is an approach to web development or a set of best practices, used to create a
> layout that can respond to the device's viewport size.
> 
> You can read more about the history of
> [Responsive Design](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Responsive_Design).
> 
> ## Container and Viewport Based Responsive Styling
> 
> We will take a look at Container-based and Viewport-based Responsive Styling, what they are and
> which to use.

Reviewer: ```suggestion
Responsive styling is an approach to web development that adjusts layout and appearance based on the
device's viewport size or an element's container size.

## What is Responsive Styling?

SHORT DESCRIPTION OF RESPONSIVE STYLING GOES HERE.

Read more about the history of
[Responsive Design](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Responsive_Design).
```

Replace the `SHORT DESCRIPTION OF RESPONSIVE STYLING GOES HERE` placeholder with an actual description.

The Canvas Site will strip out the H1 and every paragraph underneath it up until the next H2 (not including the H2). This information is typically rendered in the gray preamble section on the site instead:

<img width="823" alt="image" src="https://user-images.githubusercontent.com/146020/198417018-f11f546d-a238-46c8-9da2-6e8cbe2efabb.png">

If we want the Responsive Design MDN link to show up on the site, it needs to be underneath an H2 following the H1. I've added the `## What is Responsive Styling?` section in my suggestion to give that link a place to live -- it replaces the `## Container and Viewport Based Responsive Styling` which I don't think is needed.

> which to use.

Author: Thank you for the feedback on this James!

### Container Based Responsive Styling

> With viewport-based responsive styling (media queries), the user's viewport (user's visible area of

Author: I went ahead and removed the portion discussing viewport-based responsive styling and just discussed container-based. You're right and it was above the container-based portion before.

a web page) is measured and styles are applied based on the size of that viewport. We will do the
same with Container Based Responsive Styling but, we will just be measuring the size of the
container the element is in instead of the user's viewport.

In the example, we see that there is a `maxWidth` set on the Box component of `1000`. This will set
the maximum width of that element at `1000px`. What we can do with Container Based Responsive
Styling is measure when that container's width reaches certain breakpoints
`(i.e. 1440px, 1024px, 768px, 320px, 0)` and apply styles based on those breakpoints. If the
container's width were to reach `768px`, we can apply styles to that breakpoint.

```tsx
<Box maxWidth={1000}>Hello World</Box>

> ```

Author: Agreed

In the interactive example below, when the dropdown's value is updated, the container size will
change to match the size shown in the dropdown. This is just an example of how container-based

> styles work and will be explained more in depth below.

Author: Agreed

<ExampleCodeBlock code={ResponsiveContainer} />

> ### Applying Container-Based Responsive Styling
> 
> For container-based responsive styling, we have utilities that will allow you to apply responsive
> styles at those breakpoints. There are four utilities/hooks:
> `useResizeObserver | useResponsiveContainerStyles`
> 
> `useResizeObserver` - This is a hook that will provide the width of the container. This can be used
> to measure container size as well.
> 
> `useResponsiveContainerStyles` - This is a hook that will allow you to create container-based
> responsive styles with objects (as you can see in the example below). This hook accepts three
> arguments with the third being optional (Style Objects, Container Width, Theme). In each style
> object, there are five sizes that it will accept: `zero, small, medium, large and extra large`.
> These give sizes represent the breakpoints that are at the bottom of this page. The sizes will act
> like `min-width`. For example, if you want to apply styles from `medium` and up, then you would
> write those styles under `m`.

Reviewer: I don't think we need the `### Applying Container-Based Responsive Styling` here. Let's just get right to explaining the hooks.

```suggestion
This example uses the following hooks:

- `useResizeObserver` provides the width (and height) of the container.

- `useResponsiveContainerStyles` allows you to create container-based responsive styles with
  objects. This hook accepts three arguments: style objects, container width, and theme (optional).
  Each style object accepts keys corresponding to the following sizes: `zero`, `s`, `m`, `l` and
  `xl`. These sizes represent [Canvas breakpoints](#canvas-breakpoints). The sizes act like
  `min-width`. For example, if you want to apply styles from `medium` and up, then you would write
  those styles under `m`.
```

Author: Agreed

> In the example below, we are importing these utilities:
> `useResizeObserver | useResponsiveContainerStyles`
> 
> See example below:
> 
> ```tsx
> import {Flex, Box} from '@workday/canvas-kit-react/layout';
> import {useResizeObserver, useResponsiveContainerStyles} from '@workday/canvas-kit-react/common';
> 
> const ref = React.useRef(null);
> const [width, setWidth] = React.useState(0);
> 
> useResizeObserver({
>   ref: ref,
>   onResize: data => {
>     setWidth(data.width || 0);
>   },
> });
> 
> const containerResponsiveStyles = useResponsiveContainerStyles(
>   {
>     flex: {
>       flexDirection: 'column',
>       padding: 'm',
>       depth: 1,
>       borderRadius: 'l',
>       zero: {
>         backgroundColor: 'Red',
>       },
>       s: {
>         backgroundColor: 'Orange',
>       },
>       m: {
>         backgroundColor: 'Yellow',
>       },
>       l: {
>         backgroundColor: 'Green',
>       },
>       xl: {
>         backgroundColor: 'Blue',
>       },
>     },
>     box: {
>       padding: 's',
>     },
>   },
>   width
> );
> 
> return (
>   <Box ref={ref}>
>     <Flex {...containerResponsiveStyles.flex}>
>       <Box {...containerResponsiveStyles.box}>Hello World</Box>
>     </Flex>
>   </Box>
> );
> ```

Reviewer: Again, I don't think we need this example. It doesn't seem to add much to the interactive example above. Let me know if I'm missing something.

```suggestion
```

> ### Viewport Based Responsive Styling

Reviewer: Let's make this into an H2 as well.

```suggestion
## Viewport-based Responsive Styling
```

> The viewport is the user's visible area of a web page and it varies with screen size.

Reviewer: Wordsmithing.
```suggestion
The viewport is the user's visible area of a web page (varies with screen size).
```

We can use media queries to run a series of tests that will measure the screen size or resolution
and apply CSS styles specifically to an element.

> An example of CSS media queries is below:
> 

Reviewer: Don't need this sentence 
```suggestion
```

```css
@media screen and (min-width: 800px) {
  .container {
    margin: 1em 2em;
  }
}
```

> In the example below, viewport-based styles are being applied by the size of the user's viewport.
> This is by using the breakpoint utilities (`up(), down(), between() and only()`). At the top of the
> code you will see `up() and down()` being used inside of the `StyledHeading` component. These
> utilities will be explained more in depth below the example. The size of the browser will have to be
> adjusted to see the changes take place on the UI Layout.

Reviewer: ```suggestion
In the interactive example below, styles are applied based on the size of the user's viewport using
our breakpoint utilities: `up()`, `down()`, `between()` and `only()`. Note the use of `up()` and
`down()` near the top of the example code.

Adjust the width of your browser to see the layout adjust to the available width.
```

> <ExampleCodeBlock code={ResponsiveViewport} />

Reviewer: This example is very cool, btw. I love how the physical device is part of the example.

Author: Thank you!

> ### Applying Viewport-Based Responsive Styling

Reviewer: Let's just call this section Breakpoint Utilities.
```suggestion
### Breakpoint Utilities
```

Author: Agreed

> We have three utilities to apply viewport-based responsive styling.
> 
> `up() | down() | between() | only()`
> 
> Each has a specific role in applying viewport-based responsive styles.

Reviewer: ```suggestion
We have four utilities to apply viewport-based responsive styling: `up()`, `down()`, `between()`,
and `only()`. Each has a specific role in applying viewport-based responsive styles.
```

import {space} from '@workday/canvas-kit-react/tokens';
const theme = useTheme();
const {up} = theme.canvas.breakpoints;
const mediaQuery = up('l'); // Returns '@media (min-width: 1024px)'
const styles = {
  [mediaQuery]: {
    padding: space.m,
  },
};
```

`down()` - Returns a media query below the `max-width` for the range of a given breakpoint. Given an
`end` breakpoint key `("zero", "s", "m", "l", "xl")`, this function returns a media query (string)
using a `max-width`.

> **Note**: This function subtracts `0.5px` from the next breakpoint value to prevent collisions.
> For example, `breakpoints.values.s`, has a `min-width` of `320px`, and the `max-width` is
> `767.5px`). If the `xl` breakpoint is provided, this function returns a media query with only a
> `min-width` of `0`, as seen in the second example below. See example below.

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';
const theme = useTheme();
const {up} = theme.canvas.breakpoints;
const mediaQuery = up('l'); // Returns '@media (min-width: 1024px)'
const styles = {
  [mediaQuery]: {
    padding: space.m,
  },
};

> ```

Reviewer: This looks like an unintended duplicate of the example for `up()` above. Can you double-check this and remove it if this is the case?

Also, my eyes are starting to glaze over from reading docs and this Breakpoints Utilities section is particularly monotonous (through no fault of yours, these utilities just all look the same) 😅 . Can you verify there aren't similar issues throughout this section?

Author: This was a duplicate and should have been showing the down example. 

The utilities do all look similar ha I agree.

`min-width` and `max-width`.

> **Note**: This function subtracts `0.5px` from the next breakpoint value to prevent collisions.
> For example, `breakpoints.values.s`, has a `min-width` of `320px`, and the `max-width` is
> `767.5px`).

If the `xl` breakpoint is provided, this function returns a media query with only a `min-width` of
`1440px`, as seen in the second example below.

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';

const theme = useTheme();
const {only} = theme.canvas.breakpoints;
const mediaQuery = only('s'); // Returns '@media (min-width: 320px) and (max-width: 767.5px)'
const styles = {
  [mediaQuery]: {
    padding: space.s,
  },
};
```

This example uses the `xl` breakpoint and only adds a `min-width` of `1440px` to the media query.

```ts
import {useTheme} from '@workday/canvas-kit-react/common';
import {space} from '@workday/canvas-kit-react/tokens';

const theme = useTheme();
const {only} = theme.canvas.breakpoints;
const mediaQuery = only('xl'); // Returns '@media (min-width: 1440px)'
const styles = {
  [mediaQuery]: {
    padding: space.s,
  },
};
```

> ### Viewport vs Container

Reviewer: This section should be further up in the document. I'd consider putting it in the `## What is Responsive Styling?` section I suggested at the top.

Author: I looked at this and in terms of flow for the docs, I revamped the entire section. I added more description of container vs viewport

If you are measuring a user's viewport size, then you should be using Viewport Based Responsive
Styling.

If you only want to measure the size of a container, then you should be using Container Based
Responsive Styling.

> **Note**: A container's size can be affected by surrounding containers or the viewport size.

> ### Canvas Kit Breakpoint Sizes

Reviewer: Rename this heading to Canvas Breakpoints and also consider moving it further up. This could also go in the `## What is Responsive Styling?` section.
```suggestion
### Canvas Breakpoints
```

Author: Moved up and added to the `What is Responsive Styling?` section

import {Meta, ArgsTable} from '@storybook/addon-docs';
import {Box} from '@workday/canvas-kit-react/layout';

> //

Reviewer 2: is this supposed to be here ?

Author: The comment slashes no. I will update that.

import {
  useResizeObserver,
  ResponsiveContextProvider,
  useResponsiveStyles,
} from '@workday/canvas-kit-react/common';
import {ResponsiveContainer} from './examples/ResponsiveStyles';
import {ResponsiveViewport} from './examples/ResponsiveViewport';

<Meta title="Hooks and Utilities/Responsive Styling" />

# Responsive Styling

Responsive styling is an approach to web development or a set of best practices, used to create a

> layout that can respond to the device being used to view the content.

Reviewer 2: do we wanna say device or maybe size of the view port? or screen size?

Author: ...respond to the device's screen size that is being used. 

Thoughts?

Author: Updated. 

...respond to the device's viewport size

> layout that can respond to the device being used to view the content.

Reviewer 3: ```suggestion
layout that can respond to the viewport size.
```

Reviewer 3: This also is based on the container size right which isn't always based on the screen or viewport?

Author: Updated

You can read more about the history of
[Responsive Design](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Responsive_Design).

## Container and Viewport Based Responsive Styling

We will take a look at Container-based and Viewport-based Responsive Styling, what they are and
which to use.

### Container Based Responsive Styling

> Now, with viewport-based responsive styling (media queries), the user's viewport (user's visible

Reviewer 3: ```suggestion
With viewport-based responsive styling (media queries), the user's viewport (user's visible
```

Author: Updated

area of a web page) is measured and styles are applied based on the size of that viewport. We will
do the same with Container Based Responsive Styling but, we will just be measuring the size of the
container the element is in instead of the user's viewport.

Take the below example of a component:

```tsx
<Box maxWidth={1000}>Hello World</Box>
```

In the example above, we see that there is a `maxWidth` set on the Box component of `1000`. This
will set the maximum width of that element at `1000px`. What we can do with Container Based
Responsive Styling is measure when that container's width reaches certain breakpoints
`(i.e. 1440px, 1024px, 768px, 320px, 0)` and apply styles based on those breakpoints. If the
container's width were to reach `768px`, we can apply styles to that breakpoint.

> In the interactive example below, when the dropdown's value is updated, the container size will

Reviewer 3: in the above paragraph you mention an example above and in the bottom paragraph you mention an example below. Maybe we can just describe what is happening in one section and let the example describe what's happening. We have a similar patterns in our MDX where we describe some storybook example and then get the example below.

Author: Definitely agree, I will update this

change to match the size shown in the dropdown. This is just an example of how container-based
styles work and will be explained more in depth below.

<ExampleCodeBlock code={ResponsiveContainer} />

### Applying Container-Based Responsive Styling

For container-based responsive styling, we have utilities that will allow you to apply responsive
styles at those breakpoints. There are four utilities/hooks:

> `useResizeObserver | ResponsiveContextProvider | useResponsiveContext | useResponsiveStyles`

Reviewer 2: ```suggestion
`useResizeObserver | ResponsiveContextProvider | useResponsiveContext | useResponsiveContainerStyles`
```

Author: Updated

> `useResizeObserver` - This is a hook that will provide the width of the container. This can used to

Reviewer 3: ```suggestion
`useResizeObserver` - This is a hook that will provide the width of the container. This can be used to
```

Author: Updated

measure container size as well.

`ResponsiveContextProvider` - This React Context Provider that will accept the width of the
container from `useResizeObserver`.

`useResponsiveContext` - This hook subscribes to the `ResponsiveContextProvider` and will determine
whether the current width provided from `useResizeObserver` is within a particular theme breakpoint.

> `useResponsiveContainerStyles` - This is a hook that will allow you to create container-based

Reviewer 2: this does not match line 56

Author: Updated

responsive styles with objects (as you can see in the example below). This hook accepts three
arguments with the third being optional (Style Objects, Container Width, Theme). In each style
object, there are five sizes that it will accept: `zero, small, medium, large and extra large`.
These give sizes represent the breakpoints that are at the bottom of this page. The sizes will act
like `min-width`. For example, if you want to apply styles from `medium` and up, then you would
write those styles under `m`.

In the example below, we are importing three of those utilities:

> `useResizeObserver | ResponsiveContextProvider | useResponsiveStyles`

Reviewer 2: ```suggestion
`useResizeObserver | ResponsiveContextProvider | useResponsiveContainerStyles`
```

Author: Updated

See example below:

```tsx
import {Flex, Box} from '@workday/canvas-kit-react/layout';
import {
  useResizeObserver,
  ResponsiveContextProvider,
  useResponsiveContainerStyles,
} from '@workday/canvas-kit-react/common';

const ref = React.useRef(null);
const [width, setWidth] = React.useState(0);

useResizeObserver({
  ref: ref,
  onResize: data => {
    setWidth(data.width || 0);
  },
});

const containerResponsiveStyles = useResponsiveContainerStyles(
  {
    flex: {
      flexDirection: 'column',
      padding: 'm',
      depth: 1,
      borderRadius: 'l',
      zero: {
        backgroundColor: 'Red',
      },
      s: {
        backgroundColor: 'Orange',
      },
      m: {
        backgroundColor: 'Yellow',
      },
      l: {
        backgroundColor: 'Green',
      },
      xl: {
        backgroundColor: 'Blue',

>       },

Reviewer 2: **snaps for ROYGBIV order 🌈

    },
    box: {
      padding: 's',
    },
  },
  width
);

return (
  <ResponsiveContextProvider width={width}>
    <Box ref={ref}>
      <Flex {...containerResponsiveStyles.flex}>
        <Box {...containerResponsiveStyles.box}>Hello World</Box>
      </Flex>
    </Box>
  </ResponsiveContextProvider>
);
```

### Viewport Based Responsive Styling

> The viewport is the user's visible area of a web page and it varies with the device.

Reviewer 2: again should we say device or screen size?

Author: Screen size for sure
